docs/src/main/sphinx/develop/connectors.rst

```diff
@@ -223,10 +223,79 @@ features, like:
 Note that data modification also requires implementing
 a :ref:`connector-page-sink-provider`.
 
+When Trino receives a ``SELECT`` query, it parses it into an Intermediate
+Representation (IR). Then, during optimization, it checks if connectors
+can handle operations related to SQL clauses by calling one of the following
+methods of the ``ConnectorMetadata`` service:
+
+* ``applyLimit``
+* ``applyTopN``
+* ``applyFilter``
+* ``applyProjection``
+* ``applySample``
+* ``applyAggregation``
+* ``applyJoin``
+* ``applyTableFunction``
+* ``applyTableScanRedirect``
+
+Connectors can indicate that they don't support a particular pushdown or that
+the action had no effect by returning ``Optional.empty()``. Connectors should
+expect these methods to be called multiple times during the optimization of
+a given query.
+
+.. warning::
+
+  It's critical for connectors to return ``Optional.empty()`` if calling
+  this method has no effect for that invocation, even if the connector generally
+  supports a particular pushdown. Doing otherwise can cause the optimizer
+  to loop indefinitely.
+
+Otherwise, these methods return a result object containing the connectors
```

Comment:
> the connectors table handle

It's actually a new table handle representing the virtual table derived from applying the operation (filter, project, limit, etc) to the table produced by the table scan node.